Add LSIF support

chrismwendt opened the pull request:
**Goals**

The goal of **this PR** is to enable us in the Sourcegraph organization to upload LSIF data for the sourcegraph/sourcegraph repository to Sourcegraph.com.

The next goal (not for this PR) is to support other users uploading LSIF for their own repositories by GopherCon (July 24).

The kind of feedback that would be most helpful would be: pointing out problems with the architecture or usability. Here are some areas I'm looking for feedback on:

- Auth: I plan to restrict the `/upload` endpoint to admin access tokens, and keep the `/request` endpoint open to normal users (or anonymous users, if the instance is public) cc @sourcegraph/core-services
- API: this currently resides at `.api/lsif`, but I plan to move it under GraphQL cc @sourcegraph/core-services
- The LSIF HTTP server is Node.js: are we OK with running a Node.js process on the backend? I'm reusing a few files from https://github.com/microsoft/vscode-lsif-extension/tree/master/server (and asked about an npm package https://github.com/microsoft/vscode-lsif-extension/issues/13), but they could probably get rewritten in Go in ~1 week cc @beyang @felixfbecker @lguychard
- General architecture and anything I'm overlooking (file storage, caching, API, etc.)

**Background** https://github.com/sourcegraph/sourcegraph/issues/4692

**Overview of implementation** see lsif/README.md and lsif/server/README.md

**Changelog** Doesn't update the CHANGELOG because there's no LSIF-scope for access tokens yet, which is OK for an unannounced MVP that we try out for ourselves on the sourcegraph/sourcegraph repository https://sourcegraph.slack.com/archives/C0C324C91/p1562109046016800?thread_ts=1562105156.005000&cid=C0C324C91

**Test plan**

- cd into `sourcegraph/codeintellify`
- `lsif-tsc -p tsconfig.json --noContents --out data.lsif`
- `env SRC_ENDPOINT=http://localhost:3080 SRC_ACCESS_TOKEN=$DEV_SRC_ACCESS_TOKEN REPOSITORY=github.com/sourcegraph/codeintellify COMMIT=$(git rev-parse HEAD) bash ~/path/to/sourcegraph/sourcegraph/lsif/upload.sh data.lsif` (currently fails at the upload step because I haven't figured out how to implement auth yet)
- Make sure it says "Upload successful."
- TODO add LSIF to each language extension (basic-code-intel can be a vehicle for this)
- Open http://localhost:3080/github.com/sourcegraph/codeintellify/-/blob/src/hoverifier.ts
- You should see hovers and defs working

**TODO**

- [x] Check if LSIF data for a given repo@commit `/exists` before sending hover/def/ref requests
- [x] Check the version number during upload, return 400 unless it matches the version that the LSIF DB code expects
- [x] Restrict the `/upload` endpoint to admins (in the future, we might create a new LSIF scope for access tokens)
- [x] Write a Dockerfile, add it to gen-pipeline.go
- [x] Address TODOs in code, refactor

**Potential future ideas** (we can move select tasks into the TODOs above)

- Translate repository name to repository ID
- Add metrics to track upload rate, upload size, request rate, size of `Database` cache, etc.
- Add throttling for uploads and requests
- Restrict uploads per-repository (see discussion about this https://sourcegraph.slack.com/archives/C0C324C91/p1562105156005000)
- Add a new kind of token (different from user access tokens): LSIF upload token
- Support fetching LSIF data from external storage (e.g. Amazon S3, like Codecov does)
- Support multiple replicas of this LSIF server and figure out storage and load balancing
- Fall back to the most recent commit that does have LSIF data if the current one doesn't
- Infer repository, commit, etc. from build environment (Travis, CircleCI, Buildkite, etc.) like Codecov does

tsenart commented:
> API: this currently resides at .api/lsif, but I plan to move it under GraphQL cc @sourcegraph/core-services

What's the rationale for wanting to move it to GraphQL?

> Auth: I plan to restrict the /upload endpoint to admin access tokens, and keep the /request endpoint open to normal users (or anonymous users, if the instance is public) cc @sourcegraph/core-services

A separate scope would be great. Perhaps we could track that in a separate issue.

felixfbecker commented:
### API

Could you document what the endpoints accept? It's currently not clear to me what a request body needs to look like. What does "file upload" mean? What's the schema?
For what is LSIF content uploaded and requested, a repo, a file, a package, ...?

It also seems these are RPC ("request" and "upload" are both verbs), which seems like a missed opportunity here to make it more RESTful. It's valuable to follow HTTP semantics as much as possible (e.g. for easy caching), and it seems easy here. Why not make it something like `POST /.api/lsif/repo@rev/file` and `GET /.api/lsif/repo@rev/file`?

### Node

> The LSIF HTTP server is Node.js: are we OK with running a Node.js process on the backend?

We already have a Node microservice for badges, and I think it makes a lot of sense to use it here, given that Microsoft's LSIF stuff will be in TS. Everything else of code-nav is in TS too.

### Storage

What's the thinking behind storing the LSIF contents as files? How will we want to query it? Files seem obvious if we just want to echo the uploaded raw data, but in that case is this server just a custom-built S3? If there is any querying we want to support, we should probably save it in a database that allows us to efficiently do those queries

### GraphQL

Since LSIF is a graph GraphQL seems like a good choice if we want to support specific querying on it.

### Microsoft's package

Did you ask them to publish it? It seems like it's a VS Code extension so they probably won't do it unless we ask them. Note the files have "See License.txt in the project root for license information." in the header and are licensed under a different license than this repo so I don't think you can copy them like this PR does.

### Code

I find the code hard to review rn, could you add inline documentation? I have more feedback but I'd like to avoid back and fourth that is caused by potential misunderstandings.

### Meta

It would be nice to have this discussion in a design document instead of on a pull request for the implementation

chrismwendt commented:
@tsenart I'm guessing it's easier to apply permissions to endpoints in GraphQL than `.api/lsif`, but where would you put this?

I added a section for "Post-merge TODO" and added the LSIF scope idea to it.

@felixfbecker Great points on the design/architecture! ✨ And :+1: to a Google doc to discuss this further: https://docs.google.com/document/d/1W0L0Z34Og7j0WRHEesxs_92I0vOmy-eC0BvbOVzFGww/edit# I'll be responding and adding my own notes there, too.

chrismwendt commented:
The `http-server.ts` is ready for review :sparkles:

I'm happy to do a video chat :movie_camera: in case there are more than a handful of comments (cc @felixfbecker)

lguychard commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,353 @@
+/**
+ * Reads an integer from an environment variable or defaults to the given value.
+ */
+function readEnvInt({ key, def }: { key: string; def: number }): number {
```
```suggestion
function readEnvInt({ key, default }: { key: string; default: number }): number {
```

chrismwendt replied:
`default` is a reserved word in TypeScript. Changed to `defaultValue`.

lguychard commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,353 @@
+ * this limit if a single LSIF file is larger than this, otherwise storage will
+ * be kept under this limit.
+ */
+const softMaxStorage = readEnvInt({ key: 'SRC_LSIF_SOFT_MAX_STORAGE', def: 100 * 1024 * 1024 * 1024 })
```
Document the default in the docstring? Also, top-level globals should be upper snake case (`SOFT_MAX_STORAGE`)

lguychard commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,353 @@
+/**
+ * List of supported `Database` methods.
+ */
+const supportedMethods = ['hover', 'definitions', 'references']
```
I would make this a string union instead, with an `isSupportedMethod()` typeguard.

lguychard commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,353 @@
+                })
+            } catch (e) {
+                if ('code' in e && e.code === 'ENOENT') {
+                    res.send({ error: `No LSIF data available for ${repository}@${commit}.` })
```
Should this send a 404?

felixfbecker commented:
Please make sure the code is checked by our TSLint, ESLint and Prettier :)

chrismwendt commented on lsif/README.md:
```diff
@@ -0,0 +1,13 @@
+# LSIF code intelligence
```
I'm not sure, but I'll leave it in the top level `lsif/` unless anyone else would prefer it live elsewhere.

chrismwendt commented on lsif/server/README.md:
````diff
@@ -0,0 +1,46 @@
+For example, for `github.com/sourcegraph/codeintellify` at commit `c21c0da7b2a6cacafcbf90c85a81bf432020ad9b`:
+
+```
+base64repository:Z2l0aHViLmNvbS9zb3VyY2VncmFwaC9jb2RlaW50ZWxsaWZ5,commit:c21c0da7b2a6cacafcbf90c85a81bf432020ad9b.lsif
````
Switched to URL encoding of the repo name

chrismwendt commented on cmd/frontend/internal/httpapi/httpapi.go:
```diff
@@ -45,6 +50,15 @@ func NewHandler(m *mux.Router) http.Handler {
+		proxy := httputil.NewSingleHostReverseProxy(lsifServerURL)
+		m.Get(apirouter.LSIFUpload).Handler(trace.TraceRoute(http.HandlerFunc(proxyHandlerLSIFUpload(proxy))))
+		m.Get(apirouter.LSIF).Handler(trace.TraceRoute(http.HandlerFunc(proxyHandlerLSIF(proxy))))
+	}
```
@keegancsmith Is the LSIF-specific logic OK here, or should I move this to a different file/package? I'm not sure where these usually go.

slimsag replied:
This code here is fine and can be refactored out later if it grows larger than this amount of code.

The `proxyHandler`s should go into a separate file in this package.

chrismwendt commented:
Could I get review on these files? :memo:

- [ ] `cmd/` dir
- [ ] `dev/{Procfile,launch.sh}`
- [ ] `enterprise/dev/Procfile`
- [ ] `lsif/server/src/{Dockerfile,build.sh}`

slimsag commented on cmd/frontend/internal/httpapi/httpapi.go:
```diff
@@ -55,6 +69,24 @@ func NewHandler(m *mux.Router) http.Handler {
 	return m
 }
 
+func proxyHandlerLSIF(p *httputil.ReverseProxy) func(http.ResponseWriter, *http.Request) {
```
Suggest inverting the name to remove ambiguity about what these belong to. i.e. `proxyHandlerLSIF` -> `lsifProxyHandler`

slimsag commented on cmd/frontend/internal/httpapi/router/router.go:
```diff
@@ -121,4 +123,6 @@ func addTelemetryRoute(m *mux.Router) {
 func addGraphQLRoute(m *mux.Router) {
 	m.Path("/graphql").Methods("POST").Name(GraphQL)
+	m.Path("/lsif/upload").Methods("POST").Name(LSIFUpload)
+	m.Path("/lsif/{rest:.*}").Methods("POST").Name(LSIF)
```
It does not make sense why this would be under a function named `addGraphQLRoute`. Move it into a separate one like `addRegistryRoute` is?

slimsag replied:
Or better, just call it inline above. I don't see any good reason that these functions exist.

chrismwendt replied:
I inlined the routes. `addGraphQLRoute` is probably a function because it's called twice (once from the public API router and once from the internal one).

slimsag commented on cmd/server/Dockerfile:
```diff
@@ -44,7 +44,7 @@ RUN apk update && apk add --no-cache \
     'bash=4.4.19-r1' 'postgresql-contrib=11.4-r0' 'postgresql=11.4-r0' \
     'redis=3.2.12-r0' bind-tools ca-certificates git@edge \
-    mailcap nginx openssh-client su-exec tini
+    mailcap nginx openssh-client su-exec tini nodejs=10.14.2-r0
```
By how much does this bloat our Dockerfile, I am curious? (not a blocker)

chrismwendt replied:
The `nodejs` Alpine package takes up ~22MB:

```
$ apk -L info -s nodejs
nodejs-10.14.2-r0 installed size:
24080384

nodejs-10.14.2-r0 contains:
usr/bin/node
usr/share/systemtap/tapset/node.stp

$ du -sch /lsif-server
56.3M	/lsif-server/
```

And `node_modules/` is ~55MB (yikes!). Parcel bundles it to 17KB, so I'll do that instead of copying `node_modules/`. That will also simplify the build scripts a bit.

felixfbecker replied:
Is there no Node 12 available?

slimsag commented on cmd/server/build.sh:
```diff
@@ -51,6 +51,15 @@ done
+cp lsif/server/out/* "$OUTPUT/lsif-server"
+cp -r lsif/server/node_modules "$OUTPUT/lsif-server/node_modules"
+printf "#!/usr/bin/env bash\\nnode ../../../lsif-server/http-server.js" > "$bindir/lsif-server"
+chmod +x "$bindir/lsif-server"
```
Instead of generating this shell script, you could use the same pattern syntect_server uses in the `shared.go` `procfile` like:

```
`lsif-server: sh -c 'node ../../../lsif-server/http-server.js'`
```

slimsag commented on cmd/server/shared/shared.go:
```diff
@@ -96,6 +97,7 @@ func Main() {
 	// Next persistence
 	{
 		SetDefaultEnv("SRC_REPOS_DIR", filepath.Join(DataDir, "repos"))
+		SetDefaultEnv("SRC_LSIF_STORAGE_ROOT", filepath.Join(DataDir, "lsif"))
```
I would ditch the `SRC_` prefix, I think we only have that in others for legacy reasons.

slimsag commented on dev/Procfile:
```diff
@@ -16,3 +16,4 @@ management-console: PROJECT_ROOT=./cmd/management-console TLS_CERT=$HOME/.source
 management-console-web: cd ./cmd/management-console/web && ./serve.sh
 # jaeger: docker run --name=jaeger --rm -p5775:5775/udp -p6831:6831/udp -p6832:6832/udp -p5778:5778 -p16686:16686 -p14268:14268 jaegertracing/all-in-one:latest
 docsite: .bin/docsite -config doc/docsite.json serve -http=localhost:5080
+lsif: yarn --cwd lsif/server tsc-watch --onSuccess "node out/http-server.js" --noClear
```
I am surprised this line is not something like `cd lsif/server && yarn run`, which would make it easier to run it separately from the rest of Sourcegraph (sometimes useful for debugging).

slimsag commented on lsif/README.md:
```diff
@@ -0,0 +1,13 @@
+# LSIF code intelligence
+
+BEFOER MERGING consider moving some of this to docs now or later
```
```suggestion
BEFORE MERGING consider moving some of this to docs now or later
```

slimsag replied:
😆

slimsag commented on lsif/README.md:
```diff
@@ -0,0 +1,13 @@
+In this directory:
+
+- [upload-lsif.sh](upload-lsif.sh): a script that uploads an LSIF file to Sourcegraph (usually used in CI, similar to [Codecov's Bash Uploader](https://docs.codecov.io/docs/about-the-codecov-bash-uploader))
+- [server/](server/): an HTTP server which runs inside of Sourcegraph (for auth), receives and stores LSIF file uploads, and services requests for hovers/defs/refs
```
```suggestion
- [server/](server/): an HTTP server which runs behind Sourcegraph (for auth), receives and stores LSIF file uploads, and services requests for hovers/defs/refs
```

slimsag commented on lsif/server/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+# hadolint ignore=DL3018
+RUN apk add --no-cache nodejs=10.14.2-r0 tini
+
+COPY / /
```
This seems oddly difficult to read. I assume `/` on the host is copying the Docker build context root -- not your root FS, still it seems you could instead write:

```
COPY out /lsif-server
```

And then your run command below could become `/lsif-server/http-server.js` which seems more clear to me.

chrismwendt commented:
@slimsag Thanks, I applied all of your suggestions :+1:

chrismwendt commented:
@felixfbecker @lguychard Any more feedback on the TypeScript code blocking approval?

felixfbecker commented on lsif/extension/src/lsif.ts:
```diff
@@ -0,0 +1,161 @@
+import * as path from 'path'
+import queryString from 'query-string'
```
The native `URL`/`URLSearchParams` API can do this, no need to add this to the bundle

felixfbecker commented on lsif/extension/src/lsif.ts:
```diff
@@ -0,0 +1,161 @@
+            }),
+            body: JSON.stringify({
+                method,
+                params: [pathFromDoc(doc), params],
```
What format does this body follow? LSP uses object params and JSON RPC. However, LSIF spec defines this:

> The corresponding LSP requests have one of the following two forms:
> 
> ```ts
> request(uri, method) -> result
> request(uri, position, method) -> result
> ```

This however seems to be none of the two

chrismwendt replied:
The `params` here are for the `Database` interface method calls, not for direct inspection of the graph.

I refactored the `params` so they match the arguments to `Database` method calls.

felixfbecker commented on lsif/extension/src/lsif.ts:
```diff
@@ -0,0 +1,161 @@
+                            )
+                            .join('\n'),
+                        kind: sourcegraph.MarkupKind.Markdown,
+                    },
```
This looks like a inferior version of [`convertHover`](https://github.com/sourcegraph/lsp-client/blob/7037acebf61e427d91f0976262fd6c66209f60b0/src/lsp-conversion.ts#L32-L59). You could just import it from the library

chrismwendt replied:
Let's figure out what the dependency tree will look like later https://sourcegraph.slack.com/archives/CHXHX7XAS/p1562699201039900

felixfbecker commented on lsif/extension/src/lsif.ts:
```diff
@@ -0,0 +1,161 @@
+                if (!body) {
+                    return null
+                }
+                return body.map((definition: any) => ({ ...definition, uri: setPath(doc, definition.uri) }))
```
Why `any`? Shouldn't this be a known LSP type we can reference?

felixfbecker commented on lsif/server/README.md:
````diff
@@ -0,0 +1,46 @@
+LSIF files are stored on disk with the following naming convention:
+
+```
+urlEncodedRepository:$URL_ENCODED_REPOSITORY,commit:$40_CHAR_HASH.lsif
````
This doesn't match the shorter format in the planning doc, is that intentional?

chrismwendt replied:
Switched to the shorter format (forgot to change it earlier)

felixfbecker commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,389 @@
+/**
+ * Where on the file system to store LSIF files.
+ */
+// tslint:disable-next-line: no-string-literal
```
Why does this need to be disabled here?

chrismwendt replied:
Because `process.env.LSIF_STORAGE_ROOT` doesn't type check.

felixfbecker replied:
From a type perspective, the two are equivalent. It type checks for me

chrismwendt replied:
Huh, it type checks for me too. Not sure what was happening last time I tried.

felixfbecker commented on lsif/server/tsconfig.json:
```diff
@@ -0,0 +1,18 @@
+{
+  "compilerOptions": {
```
Extend our shared tsconfig.json

felixfbecker commented on lsif/extension/tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "extends": "./node_modules/@sourcegraph/tsconfig/tsconfig.json",
```
```suggestion
  "extends": "@sourcegraph/tsconfig",
```

felixfbecker commented on lsif/server/package.json:
```diff
@@ -0,0 +1,38 @@
+  "license": "MIT",
+  "version": "0.1.0",
+  "engines": {
+    "node": "*"
```
```suggestion
    "node": ">=10.0.0"
```
if that is the requirement?

chrismwendt replied:
Node.js 8 works

chrismwendt commented:
Thanks for the reviews, @felixfbecker I applied all of your suggestions :+1:

felixfbecker commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,394 @@
+                        default:
+                            throw new Error(`Unknown method ${method}`)
+                    }
+                    res.header('content-type', 'application/json').send(result || 'null')
```
You can use `res.json()` for this

felixfbecker commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,394 @@
+                res.send(Boolean((await createDB({ repository, commit })).stat(file)))
+            } catch (e) {
+                if ('code' in e && e.code === 'ENOENT') {
+                    res.send(false)
```
will this send `Content-Type: application/json`?

chrismwendt replied:
Yes, this sends `Content-Type: application/json`

felixfbecker commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,394 @@
+            checkRepository(repository)
+            checkCommit(commit)
+
+            if (req.header('Content-Length') && parseInt(req.header('Content-Length') || '', 10) > MAX_FILE_SIZE) {
```
The `|| ''` can't happen

chrismwendt replied:
`header()` returns `undefined` when the header doesn't exist.

felixfbecker commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,394 @@
+ * hover/defs/refs requests.
+ */
+function main(): void {
+    const app = express()
```
we may want to add a request logger like morgan too

chrismwendt commented on lsif/server/src/http-server.ts:
```diff
@@ -0,0 +1,389 @@
+    const app = express()
+
+    // This limit only applies to JSON requests (i.e. the /request endpoint).
+    app.use(bodyParser.json({ limit: '1mb' }))
```
`bodyParser.json` is only applied on the `/request` endpoint (on a newer version of this PR).
